server/ServerFSM.cpp

```diff
@@ -762,11 +770,22 @@ void MPLobby::ValidateClientLimits() {
         max_human = min_human;
         GetOptionsDB().Set<int>("network.server.human.max", max_human);
     }
+    int min_connected = GetOptionsDB().Get<int>("network.server.conn-human.min");
+    int max_disconnected = GetOptionsDB().Get<int>("network.server.disconn-human.max");
 
     // restrict minimun number of human and ai players
     if (human_count < min_human) {
         m_lobby_data->m_start_locked = true;
         m_lobby_data->m_start_lock_cause = UserStringNop("ERROR_NOT_ENOUGH_HUMAN_PLAYERS");
+    } else if (human_connected_count < min_connected) {
+        m_lobby_data->m_start_locked = true;
+        m_lobby_data->m_start_lock_cause = UserStringNop("ERROR_NOT_ENOUGH_CONNECTED_HUMAN_PLAYERS");
+    } else if (max_disconnected > 0 &&
+        !m_lobby_data->m_new_game &&
+        static_cast<int>(m_lobby_data->m_save_game_empire_data.size()) - ai_count - human_connected_count >= max_disconnected)
+    {
+        m_lobby_data->m_start_locked = true;
+        m_lobby_data->m_start_lock_cause = UserStringNop("ERROR_TOO_MANY_DISCONNECTED_HUMAN_PLAYERS");
```

Comment:
What does this mean / do in the lobby? I sort of understand stopping the server if humans disconnect from a game in progress, but why would disconnections from the lobby matter? Or when is this number reset...?

Reply:
It is to prevent violations of restrictions. Without this check it will be possible to load the game so there will be more disconnected players that allowed in options at start.

Reply:
I don't understand at all. What is a "disconnected player" when in the lobby? What is a "disconnected player" at all? A player that was connected and then disconnects, or a player slot of some sort that doesn't currently have a player in it? If the latter, then something needs to be renamed, as the meaning isn't clear.

Reply:
> What is a "disconnected player" when in the lobby? What is a "disconnected player" at all?
Each non-eliminated empire without connected player considered to be one. May be better to rename it "disconnected empire" or so?
> or a player slot of some sort that doesn't currently have a player in it
Yes, it is.

Reply:
The word "disconnected" is the problem; to me it means "was previously connected but isn't now", but you're using it to mean "doesn't have a connected player".

A better word would be "unconnected".

Also, preferably add "human" and/or "players" to the variable and option names, so
````
    int min_connected_human_empire_players = GetOptionsDB().Get<int>("network.server.conn-human-empire-players.min");
    int max_unconnected_human_empire_players = GetOptionsDB().Get<int>("network.server.unconn-human-empire-players.max");
````

Reply:
Renamed.